presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownAggregation.java

```diff
@@ -0,0 +1,118 @@
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class DereferencePushDownAggregation
+        extends DereferencePushDownRule<AggregationNode>
+{
+    public DereferencePushDownAggregation(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, aggregation());
+    }
+
+    @Override
+    protected Result pushDownDereferences(Context context, AggregationNode aggregationNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = aggregationNode.getOutputSymbols();
+        Map<Symbol, Expression> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> !isGroupingKey(entry.getKey(), aggregationNode))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+        if (pushdownExpressions.isEmpty()) {
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
+        for (Aggregation aggregation : aggregationNode.getAggregations().values()) {
+            expressionBuilder.addAll(aggregation.getCall().getArguments());
+        }
+        List<DereferenceExpression> aggregations = expressionBuilder.build().stream()
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .collect(toImmutableList());
+
+        Map<Symbol, DereferenceExpression> aggregationDereferences = aggregations.stream()
+                .distinct()
+                .filter(expression -> !expressions.containsKey(expression))
+                .collect(toImmutableMap(expression -> newSymbol(expression, context, metadata, sqlParser), Function.identity()));
+
+        List<Symbol> outputs = outputSymbols.stream()
+                .filter(symbol -> assignments.getMap().containsKey(symbol))
+                .collect(toImmutableList());
+
+        Assignments newAssignments = Assignments.builder()
+                .putAll(pushdownExpressions)
+                .putIdentities(outputs)
+                .putAll(aggregationDereferences)
+                .build();
+        ProjectNode child = new ProjectNode(context.getIdAllocator().getNextId(), aggregationNode.getSource(), newAssignments);
+
+        ImmutableMap.Builder<DereferenceExpression, Symbol> constraintBuilder = ImmutableMap.builder();
```

Comment:
The name `constraint` doesn't seem to match the actual value? Perhaps, `combinedDereferences`?